### PR TITLE
V4 doc

### PR DESCRIPTION
**Pull request by Li-Scottech**
update the docs for v4:
tutorials for VAE

### CODE REVIEW

File: docs/tutorials/vae.rst [hunk 1]
```diff
@@ -4,7 +4,7 @@ Variational Autoencoders: Step by Step
 **Variational AutoEncoders** (VAE) :cite:`vae-kingma2013auto` is one of the
 most widely used deep generative models. In this tutorial, we show how to
 implement VAE in ZhuSuan step by step. The full script is at
-`examples/variational_autoencoders/vae.py <https://github.com/thu-ml/zhusuan/blob/master/examples/variational_autoencoders/vae.py>`_.
+`examples/variational_autoencoders/vae.py <https://github.com/thu-ml/zhusuan/blob/v4/examples/variational_autoencoders/vae.py>`_.
```
Comment (thjashin) [inline docs/tutorials/vae.rst]: links should point to master because we'll eventually merge

File: docs/tutorials/vae.rst [hunk 2]
```diff
@@ -29,32 +29,31 @@ Build the Model
 ---------------
 
 In ZhuSuan, a model is constructed under a
-:class:`~zhusuan.model.base.BayesianNet` context, which enables transparent
+:class:`~zhusuan.framework.bn.BayesianNet` context, which enables transparent
```
Comment (thjashin) [inline docs/tutorials/vae.rst]: now it is not a context? maybe "constructed using :class:`~zhusuan.framework.bn.BayesianNet`", and have you make sure all these internal reference links worked?

File: docs/tutorials/vae.rst [hunk 3]
```diff
@@ -29,32 +29,31 @@ Build the Model
...
-        z = zs.Normal('z', z_mean, std=1., group_ndims=1)
+    bn = zs.BayesianNet()
+    # z ~ N(z|0, I)
+    z_mean = tf.zeros([n, z_dim])
+    z = bn.normal("z", z_mean, std=1., group_ndims=1, n_samples=n_particles)
 
 The shape of ``z_mean`` is ``[n, z_dim]``, which means that
 we have ``[n, z_dim]`` independent inputs fed into the univariate
-:class:`~zhusuan.model.stochastic.Normal` StochasticTensor. Because
+:class:`~zhusuan.distributions.univariate.Normal` StochasticTensor. Because
```
Comment (thjashin) [inline docs/tutorials/vae.rst]: you referred to distribution instead of StochasticTensor (which is in bn)

File: docs/tutorials/vae.rst [hunk 4]
```diff
@@ -127,50 +124,38 @@ another computation graph from scratch and reuse the Variables (weights here).
 If there are many stochastic nodes in the model, this process will be really
 painful.
 
-**ZhuSuan has a novel solution for this.** To observe any stochastic nodes,
-pass a dictionary mapping of ``(name, Tensor)`` pairs when constructing
-:class:`~zhusuan.model.base.BayesianNet`. This will assign observed values
-to corresponding ``StochasticTensor`` s. For example, to observe
-a batch of images ``x_batch``, write::
+**ZhuSuan has a novel solution for this.** To observe any stochastic nodes,
```
Comment (thjashin) [inline docs/tutorials/vae.rst]: following the old version, we can write (make sure the internal reference link works) "However, we usually need to pass different observations to the same `BayesianNet` more than once. To achieve this, ZhuSuan provides a new class called `MetaBayesianNet` to represent the meta version of `BayesianNet` which can repeatly produce `BayesianNet` objects by accepting different observations. The recommended way to construct a `MetaBayesianNet` is by the ... decorator. 

and refer them back to the basic concepts tutorial for more information (though it haven't been updated, let's do that later)

File: docs/tutorials/vae.rst [hunk 5]
```diff
@@ -127,50 +124,38 @@ another computation graph from scratch and reuse the Variables (weights here).
...
-the observed value, i.e., ``x_batch``, instead of the sampled tensor.
+        return bn
 
 .. note::
-
-    The observation passed must have the same type and shape as the
-    `StochasticTensor`.
-
-..
    With the help of both the ``BayesianNet`` context and factory pattern
```
Comment (thjashin) [inline docs/tutorials/vae.rst]: it's not a context anymore so i don't understand this.

File: docs/tutorials/vae.rst [hunk 6]
```diff
@@ -403,20 +387,25 @@ images to disk. Keep watching them and have fun :)
...
-            print('Epoch {}: Lower bound = {}'.format(
-                  epoch, np.mean(lbs)))
 
             if epoch % save_freq == 0:
-                images = sess.run(x_gen)
-                name = "results/vae/vae.epoch.{}.png".format(epoch)
+                images = sess.run(x_gen, feed_dict={n: 100, n_particles: 1})
+                name = os.path.join(result_path,
+                                    "vae.epoch.{}.png".format(epoch))
                 save_image_collections(images, name)
```
Comment (thjashin) [inline docs/tutorials/vae.rst]: We can show some generated image in this doc

File: docs/tutorials/vae.rst [hunk 7]
```diff
@@ -127,50 +124,38 @@ another computation graph from scratch and reuse the Variables (weights here).
...
+**ZhuSuan has a novel solution for this.** To observe any stochastic nodes, 
+we can simply apply the function: ``observe`` of :class:`~zhusuan.framework.MetaBayesianNet` ，which a dict of observed nodes.
+However, in most circumstances we would like to use 
+:class:`~zhusuan.framework.BayesianNet`  rather than
+:class:`~zhusuan.framework.MetaBayesianNet` .
+Therefore, a common practice in ZhuSuan is to use a decorator of 
+:class:`~zhusuan.framework.MetaBayesianNet` , like this::
 
-    with zs.BayesianNet(observed={'x': x_batch}):
+    @zs.meta_bayesian_net(scope="gen", reuse_variables=True)
```
Comment (thjashin) [inline docs/tutorials/vae.rst]: we need to explain the two arguments here, what's scope and reuse_variables

File: docs/tutorials/vae.rst [hunk 8]
```diff
@@ -127,50 +124,38 @@ another computation graph from scratch and reuse the Variables (weights here).
...
+
+    meta_model = build_gen(x_dim, z_dim, n, n_particles)
+    x_mean = meta_model.observe()["x_mean"]
 
 Each time the function is called, a different observation assignment can be
-passed. One may ask how to **reuse tensorflow variables** created in this
-function. ZhuSuan provides an very easy way to achieve this, that is, without
-careful management of variable scopes, one could just add a decorator to the
-function: ``@zs.reuse(scope)``, as shown in the above code. Then this function
+passed. One could also add a decorator to the function: ``@zs.reuse_variables(scope)``
```
Comment (thjashin) [inline docs/tutorials/vae.rst]: this paragraph should be in the inference section, where the code is about

File: docs/tutorials/vae.rst [hunk 9]
```diff
@@ -127,50 +124,38 @@ another computation graph from scratch and reuse the Variables (weights here).
...
-            x_logits = layers.fully_connected(lx_z, n_x,
-                                              activation_fn=None)
-            x = zs.Bernoulli('x', x_logits, group_ndims=1)
-        return model
+so that we can observe stochastic nodes in this way::
+
+    meta_model = build_gen(x_dim, z_dim, n, n_particles)
+    x_mean = meta_model.observe()["x_mean"]
 
 Each time the function is called, a different observation assignment can be
```
Comment (thjashin) [inline docs/tutorials/vae.rst]: -> Here to generate random images, we don't observe any node but just sample from them. so no argument is passed.

File: docs/tutorials/vae.rst [hunk 10]
```diff
@@ -371,25 +350,30 @@ Generate Images
...
+        x_logits = tf.layers.dense(h, x_dim)
             ...
-            x_logits = layers.fully_connected(lx_z, x_dim, activation_fn=None)
-            x = zs.Bernoulli('x', x_logits, group_ndims=1)
-        # before change: return model
-        return model, x_logits
 
-Then we add a sigmoid function to it to get a "mean" image.
+Then we add a sigmoid function to it to get a "mean" image. 
+After that, we use the function: ``output`` to provide the value ``x_mean``,
```
Comment (thjashin) [inline docs/tutorials/vae.rst]: we use `bn.output` to label this Tensor with name "x_mean" so that we can fetch it later using bn["x_mean"]

File: docs/tutorials/vae.rst [hunk 11]
```diff
@@ -371,25 +350,30 @@ Generate Images
...
+so that we can easily access it by using the function: ``observe``.
```
Comment (thjashin) [inline docs/tutorials/vae.rst]: we don't access it by observe, we access it by bn["x_mean"], but to get bn, we need to first cal meta_bn.observe.
